### PR TITLE
Refactor client and re-organize methods based on resource

### PR DESCRIPTION
**Pull request by florencelau**
Refactor client and add sorbet type checking to all methods

### CODE REVIEW

**Comment by florencelau** [comment]
@tannerljohnson @mquan looking for some feedback - is this what you had in mind for the refactor?

File: spec/greenhouse_api/client_spec.rb [hunk 1]
```diff
@@ -80,7 +80,7 @@
   end
 
   describe '#list_candidates' do
-    subject(:list_candidates) { client.list_candidates(params) }
+    subject(:list_candidates) { client.candidates.list_candidates(params) }
```
Comment (florencelau) [inline spec/greenhouse_api/client_spec.rb:83]: Not sure if I'm calling this incorrectly - it keeps saying `NameError: uninitialized constant` when I run the specs and I can't figure out why

File: lib/greenhouse_api/client.rb [hunk 2]
```diff
@@ -28,55 +28,51 @@ def initialize(api_key)
       @api_key = api_key
     end
 
-    def list_candidates(params = {})
-      list_many('candidates', params)
+    def candidates
+      Resources::Candidates
```
Comment (mquan) [inline lib/greenhouse_api/client.rb:32]: I'd imagine this results in `client.candidates.list_candidates` when using it, which is awesome. It's slightly different from what I'd imagine the implementation would look like, in that the client would return an instance of the `Resources::Candidates` and `list_candidates` is an instance method. But your current approach doesn't seem to present any issue so feel free to keep it, we can always swap it out behind the scene.

**Review by mquan — COMMENTED** [review]
@florencelau This looks awesome 👏 💯 , I just have a few minor non-blocking questions

File: spec/greenhouse_api/client_spec.rb (hunk 1 above)
Comment (mquan) [reply spec/greenhouse_api/client_spec.rb:83]: Related to the circular reference comment above, I believe we need to require the new classes prior to using them.

File: lib/greenhouse_api/client.rb (hunk 2 above)
Comment (tannerljohnson) [reply lib/greenhouse_api/client.rb:32]: I like the modularity!
I think I would have expected to be able to consume the client this way
```ruby
client.list_candidates(...)
client.get_current_offer_for_application(...)
client.list_many(...)
client.create_user(...)
```

do you think it's possible to keep the instance method pattern (without going through the resource) while still preserving the separate class structure?

for example, if I call `client.list_candidates`, do you think that could be
```ruby
def list_candidates
    Resources::Candidates.list_candidates
end
```
? 

just trying to see how we can make life a little easier for the consumer 🤔

Comment (florencelau) [reply lib/greenhouse_api/client.rb:32]: ooo yes good suggestion!

File: lib/greenhouse_api/resources/candidates.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# typed: true
+# frozen_string_literal: true
+
+require 'sorbet-runtime'
+
+module GreenhouseApi
+  module Resources
+    class Candidates < Client
```
[the remark this answers is not in the document]
Comment (florencelau) [reply lib/greenhouse_api/resources/candidates.rb]: I've tried adding `require './base_client'` in `client.rb` but it says it cannot load such a file

File: lib/greenhouse_api/resources/users.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+module GreenhouseApi
+  module Resources
+    class Users < Client
+      extend T::Sig
+
+      sig { params(first_name: String, last_name: String, email: String, on_behalf_of_id: String, additional_args: T.any(T::Boolean, String)).returns(Response) }
+      def self.create_user(first_name:, last_name:, email:, on_behalf_of_id:, **additional_args)
+        body = { first_name: first_name, last_name: last_name, email: email }.merge(additional_args).to_json
+        response = request(
+            http_method: :post,
```
[the remark this answers is not in the document]
Comment (florencelau) [reply lib/greenhouse_api/resources/users.rb]: ah must've happened when i copy and pasted, fixed!

File: lib/greenhouse_api/client.rb [hunk 5]
```diff
@@ -1,169 +1,52 @@
...
       @api_key = api_key
     end
 
+    def list_many(resource, params = {})
+      base_client.list_many(resource, params)
+    end
+    
     def list_candidates(params = {})
-      list_many('candidates', params)
+      candidates_client.list_all(params)
```
Comment (mquan) [inline lib/greenhouse_api/client.rb:15]: I kinda like your previous namespaced approach `client.candidates.all` but it also introduces a breaking change, so feel free to keep this flatten interface approach. No strong opinion from me

**Review by mquan — COMMENTED** [review]
Overall looks great, I think we should add unit tests for the new resource clients.

> I've tried adding require './base_client' in client.rb but it says it cannot load such a file

Have you tried using `require_relative './base_client'`

**Review by mquan — APPROVED** [review]
Looks great, we should also bump the version to publish it

File: lib/greenhouse_api/client.rb [hunk 6]
```diff
@@ -1,169 +1,59 @@
...
-
     def initialize(api_key)
       @api_key = api_key
     end
 
+    def list_many(resource, params = {})
+      base_client.list_many(resource, params)
+    end
+    
     def list_candidates(params = {})
```
Comment (tannerljohnson) [inline lib/greenhouse_api/client.rb:19]: nice. so it looks like this isn't going to introduce any breaking changes, right?
i see that a consumer could alternatively instantiate a candidate client

**Review by tannerljohnson — APPROVED** [review]

File: lib/greenhouse_api/client.rb (hunk 6 above)
Comment (florencelau) [reply lib/greenhouse_api/client.rb:19]: Yup, they can either instantiate a client or specific client (i.e. candidates), so no breaking changes

**Pull request merged**